types/lib/linear.go

```diff
@@ -0,0 +1,306 @@
+package lib
+
+import (
+	"bytes"
+	"fmt"
+	"strconv"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	wire "github.com/cosmos/cosmos-sdk/wire"
+)
+
+// Linear defines a primitive mapper type
+type Linear struct {
+	cdc   *wire.Codec
+	store sdk.KVStore
+	opt   *LinearOptions
+}
+
+// LinearOptions defines options for the key bytes
+type LinearOptions struct {
+	LengthKey []byte
+	ElemKey   []byte
+	TopKey    []byte
+
+	// once true forever true
+	initialized bool
+}
+
+// Should never be modified
+var cachedDefaultLinearOptions = DefaultLinearOptions()
+
+// DefaultLinearOptions returns the default setting of LinearOption
+func DefaultLinearOptions() *LinearOptions {
+	opt := LinearOptions{
+		LengthKey: []byte{0x00},
+		ElemKey:   []byte{0x01},
+		TopKey:    []byte{0x02},
+
+		initialized: true,
+	}
+
+	return &opt
+}
+
+// opt.initialized must be checked before calling init
+func (opt *LinearOptions) init() {
+	if opt == nil {
+		opt = DefaultLinearOptions()
+		return
+	}
+
+	i := uint8(0x00)
+
+	if opt.LengthKey == nil {
+		for ; ; i++ {
+			if bytes.Equal(opt.ElemKey, []byte{i}) || bytes.Equal(opt.TopKey, []byte{i}) {
+				continue
+			}
+			opt.LengthKey = []byte{i}
+			break
+		}
+	}
+
+	if opt.ElemKey == nil {
+		for ; ; i++ {
+			if bytes.Equal(opt.LengthKey, []byte{i}) || bytes.Equal(opt.TopKey, []byte{i}) {
+				continue
+			}
+			opt.ElemKey = []byte{i}
+			break
+		}
+	}
+
+	if opt.TopKey == nil {
+		for ; ; i++ {
+			if bytes.Equal(opt.LengthKey, []byte{i}) || bytes.Equal(opt.ElemKey, []byte{i}) {
+				continue
+			}
+			opt.TopKey = []byte{i}
+			break
+		}
+	}
+
+	opt.initialized = true
+}
+
+// NewLinear constructs new Linear
+func NewLinear(cdc *wire.Codec, store sdk.KVStore, opts ...*LinearOptions) Linear {
+	var opt *LinearOptions
+	switch len(opts) {
+	case 0:
+		opt = cachedDefaultLinearOptions
+	case 1:
+		opt = opts[0]
```

Comment:
Why do we take a list of `LinearOptions` if we only ever use one? I also don't quite understand why `LinearOptions` needs to be a struct, is there a reason we don't want to pass the options directly to `NewLinear` and have a `NewLinearWithDefaultOptions` helper?

Reply:
Just to make it optional. The caller can omit the options argument if they want to use the default setting. This method is applied [rational.go](https://github.com/cosmos/cosmos-sdk/blob/develop/types/rational.go)